### PR TITLE
Add concrete examples for events with stateful and stateless components

### LINKED ISSUE #7: Event section is confusing
I am reading the Event section again and again and to be honest I am more an more lost :sweat_smile: 

The purpose of this issue is to list what I find strange in the Event section in order to share my feeling before attacking and validating a possible rewrite. I tried to take the point of view of a beginner and I tried to forget how to use events in Surface.

_\<Mode Beginner On\>_

### Purpose of the default value
This section is related to the one below, but I was wondering what type should be the "default" option? String? Function?

### Use cases
The use cases of the `prop click, :event` are not clear. We use a quick example that shows how to declare a prop event but we do not know if the event is internal to the component or if we should pass a function to that prop.

```elixir
  defmodule MyButton do
    use Surface.Component

    @doc "Triggers on click"
    prop click, :event

    def render(assigns) do
      ...
    end
  end
```

This confusion is accentuated the next examples: _Handling events in Surface/LiveView_. In these examples, a prop event has not been declared but some `handle_event` functions have been declared. Moreover, reader also finds this sentence: "always declare event properties as :event" but it is not the case in the example. :confused:

Some questions raise in my mind:

- Should we use both?
- If so, when should I declare a prop event?
- Are these use cases not the same one?
- Are the name used  in `:on-{name}` the same as the one used in `prop {name}, :event`? 

### Use of the "click" event only

I suggest to add examples with other events (at least one): `keyup`, `keydown`, `debounce`, ...

`phx-click` is often used with `phx-value`:

```html
<div phx-click="inc" phx-value-myvar1="val1" phx-value-myvar2="val2">
```
will send the following map of params to the server ([doc](https://hexdocs.pm/phoenix_live_view/bindings.html#click-events)):
```elixir
def handle_event("inc", %{"myvar1" => "val1", "myvar2" => "val2"}, socket) do
```

In that case, it's not clear in Surface if I should use `phx-value-myvar` or `:on-value-my-var`. The last one seems strange to me.

_\<Mode Beginner Off\>_

I am wondering if we should rewrite this section to be more explicit and less confusing? For example, we could base the rewriting on the examples in the tests and improve them to show the "flow":

```elixir
defmodule Button do
    use Surface.LiveComponent

    prop click, :event, default: "click"

    def render(assigns) do
      ~H"""
      <button :on-click={{ @click }}>Click me!</button>
      """
    end

    def handle_event("click", _, socket) do
      {:noreply, socket}
    end
  end

  defmodule Panel do
    use Surface.LiveComponent

    prop buttonClick, :event, default: "click"

    def render(assigns) do
      ~H"""
      <div>
        <Button id="button_id" click={{ @buttonClick }}/>
      </div>
      """
    end

    def handle_event("click", _, socket) do
      {:noreply, socket}
    end
  end

  # Example without default value
  defmodule View do
    use Surface.LiveView

    def render(assigns) do
      ~H"""
      <div>
        <Panel id="panel_id" buttonClick="click"/>
      </div>
      """
    end

    def handle_event("click", _, socket) do
      {:noreply, socket}
    end
  end

  # Example with default value
  defmodule View do
    use Surface.LiveView

    def render(assigns) do
      ~H"""
      <div>
        <Panel id="panel_id" />
      </div>
      """
    end
  end
```

I believe we should highlight some points:

- the name of the prop is not correlated to the Phoenix Event name
- the list of phoenix live view events is fixed, you cannot create new event
- the prop event could have a default value that is the first argument of the `handle_event` method you would like to use
- the event you want to pass through the prop follow the same assumption
- you can pass event in nested components
- show an example with a component that has multiple events (cancel and add buttons for example)
- if the event is an "internal" event to the component, you do not have to declare `prop name, :event`

Feel free to comment and add your feeling about this section. Even if my English level is not perfect, I would be glad to take time to rewrite this section the best I can.

### PR DESCRIPTION
**Pull request by Malian**
Fixes #7

### CODE REVIEW

**Comment by Malian** [comment]
@msaraiva which tool do you use to generate the code sections? I mean the code with colours.

**Comment by msaraiva** [comment]
It depends :)

For the code in the components examples like https://surface-ui.org/uicomponents/Button, we use [Prism](https://prismjs.com/), setting the language to `JSX`. It works pretty well if the code is just `HTML/Surface`. 

For the catalogue, we use `makeup` as in [here](https://github.com/surface-ui/surface_catalogue/blob/c476d49d8d9e4edba52f1119410f0e0d082e7687/lib/surface/catalogue/components/state_dialog.ex#L150). It works great for elixir code but it will not highlight `~H"""` content. 

And finally, if you need to highlight both, Elixir and Surface code, as in https://surface-ui.org/components_basics, the only way I could find is to open the code on VS Code, copy it, and paste it inside gmail. Then I copy the resulting HTML with the inspector. Maybe there's already a plugin for VS Code that allows us to copy the underlying HTML directly from it instead of having to paste it in gmail. A couple of years ago there wasn't.

**Comment by msaraiva** [comment]
@Malian that would be great but I'm afraid it will take much longer to get it done as we'll have to do it from scratch. I managed to have a very nice working solution for Prism by just adjusting the existing JSX plugin. BTW,  I already integrated the solution with `<#Markdown>` so all you'll have to do is:

````
```elixir

you code...

```
````

Using `elixir` will already support highlighting code inside `~H"""`.

The following languages will be enabled for the website: `elixir`, `surface` (for just markup), `js` and `css`. We can add more later if we need.

**Comment by Malian** [comment]
@msaraiva I can't wait to be able to use it! We could also create a component like the `<#Example></Example>` component that just shows the highlighted code of the component. It will help us to have sync code with doc.

**Comment by msaraiva** [comment]
> We could also create a component like the <#Example></Example> component that just shows the highlighted code of the component. It will help us to have sync code with doc.

@Malian I did that already :)

Here's an example:

```
  <#Code
    language="elixir"
    module={{ SurfaceSiteWeb.GettingStarted.BindingSlotsToGenerators.Example }}
    line_range={{ 2..38 }}
    show_line_numbers
    selected_lines={{ [15, 29..30] }}
  />
```

As you can see, It also allows showing the line numbers and selecting line ranges so we can improve the docs by mentioning parts of the presented code:

image

File: lib/surface_site_web/live/events.ex [hunk 1]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
                 As you can see, we didn't have to define `phx-target` for any of the buttons. Sweet!
 
+                > **Note**: The complete list of available events, as well as other types of bindings, can be found at
+                section [Bindings](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#module-bindings)
+                in the Phoenix LiveView's docs.
+
+              </#Markdown>
+
+              <#Markdown>
+
```
Comment (msaraiva) [inline lib/surface_site_web/live/events.ex:118]:
```suggestion
```

File: lib/surface_site_web/live/events.ex [hunk 2]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                In the above examples the events have been handled by the component itself.
+                Sometime the component does not know how to handle an event and you would like to handle this event in the parent component.
+                For that kind of use case, you must declare the event in the child component by using the `prop` macro defining the type as `:event`,
+                and pass the value of that prop to the `:on-[event]` directive.
+              </#Markdown>
+
+                See the properties <LiveRedirect label="Event" to="/properties#event-property" /> section for more details about event properties.
+
+              <#Markdown>
+
```
Comment (msaraiva) [inline lib/surface_site_web/live/events.ex:136]:
```suggestion
```

File: lib/surface_site_web/live/events.ex [hunk 3]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                ### Stateless component
+                The simplest case you need to pass an event is when you create a stateless component that includes an element
```
Comment (msaraiva) [inline lib/surface_site_web/live/events.ex:138]:
```suggestion

                The simplest case you need to pass an event is when you create a stateless component that includes an element
```

File: lib/surface_site_web/live/events.ex [hunk 4]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                />
+                </footer>
+              </div>
+
+              <#Markdown>
+                Remember that stateless components cannot handle events and do not have state.
+                Events can only be handled in a `LiveView` or `LiveComponent` so we will store the state in that kind of component.
+              </#Markdown>
+
+              <#Markdown>
```
Comment (msaraiva) [inline lib/surface_site_web/live/events.ex:187]:
```suggestion

```

File: lib/surface_site_web/live/events.ex [hunk 5]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                ### Stateful component
+                In some case, you want to have a default behaviour that is handled by the component itself and you want to let the developer overrides the default implementation with a custom one. To implement a default behaviour, the component must implement an `handle_event/3` function, and so it requires to be stateful.
```
Comment (msaraiva) [inline lib/surface_site_web/live/events.ex:189]:
```suggestion

                In some cases, you may want to have a default behaviour that is handled by the component itself and let the developer overrides the default implementation with a custom one. To implement a default behaviour, the component must implement an `handle_event/3` function, and so it requires to be stateful.
```

File: lib/surface_site_web/live/events.ex [hunk 6]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                    module={{ SurfaceSiteWeb.Events.DialogExample }}
+                    line_range={{ 2..48 }}
+                    show_line_numbers
+                    selected_lines="8, 44-46"
+                  />
+                </footer>
+              </div>
+
+              <#Markdown>
+                The component implement a default `handle_event` that handles  the `close` event, and two public API functions,
```
Comment (msaraiva) [inline lib/surface_site_web/live/events.ex:216]:
```suggestion
                The component implements a default `handle_event` that handles  the `close` event, and two public API functions
```

**Review by msaraiva — COMMENTED** [review]
Hey @Malian!

I loved the changes ❤️ 

@lnr0626 and @paulstatezny feel free to add any consideration for this one too. Cheers.

File: lib/surface_site_web/live/events.ex [hunk 7]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                  />
                 </footer>
               </div>
 
               <#Markdown>
                 As you can see, we didn't have to define `phx-target` for any of the buttons. Sweet!
 
+                > **Note**: The complete list of available events, as well as other types of bindings, can be found at
+                section [Bindings](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#module-bindings)
+                in the Phoenix LiveView's docs.
```
Comment (lnr0626) [inline lib/surface_site_web/live/events.ex:113]:
```suggestion
                > **Note**: The complete list of available events, as well as other types of bindings, can be found in the
                [Bindings](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#module-bindings)
                section of the docs for Phoenix LiveView.
```

File: lib/surface_site_web/live/events.ex [hunk 8]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                    selected_lines="5, 12"
+                  />
+                </footer>
+              </div>
+
+              <#Markdown>
+              We declared a required `click` event property that we use on the `<button>` tag with the `:on-click` directive.
+
+              Now let's see how to define and pass events to that stateless component.
+              We will use twice the `Button` component with two different handling functions that have been defined in a parent live component.
```
Comment (lnr0626) [inline lib/surface_site_web/live/events.ex:162]:
```suggestion
              We will use the `Button` component twice, each with a different handling function that has been defined in a parent live component.
```

File: lib/surface_site_web/live/events.ex (hunk 5 above)
Comment (lnr0626) [reply lib/surface_site_web/live/events.ex:189]: Don't think I can update your suggestion, two small additional suggestions:

- overrides -> override
- it requires to be stateful -> it must be stateful

File: lib/surface_site_web/live/events.ex [hunk 9]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                Events can only be handled in a `LiveView` or `LiveComponent` so we will store the state in that kind of component.
+              </#Markdown>
+
+              <#Markdown>
+                ### Stateful component
+                In some case, you want to have a default behaviour that is handled by the component itself and you want to let the developer overrides the default implementation with a custom one. To implement a default behaviour, the component must implement an `handle_event/3` function, and so it requires to be stateful.
+
+                One example is a generic stateful `Dialog` component with a close button.
+                By default, if the user clicks the close button, that will close the modal. However, if you're using the dialog to show a form
+                that the user must fill in lots of information, you might want to confirm if the user really wants to close it and ask for confirmation.
```
Comment (lnr0626) [inline lib/surface_site_web/live/events.ex:193]:
```suggestion
                that the user must fill in lots of information, and you may want to ask for confirmation that the user really wants to close the dialog.
```

File: lib/surface_site_web/live/events.ex [hunk 10]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                    line_range={{ 2..48 }}
+                    show_line_numbers
+                    selected_lines="8, 44-46"
+                  />
+                </footer>
+              </div>
+
+              <#Markdown>
+                The component implement a default `handle_event` that handles  the `close` event, and two public API functions,
+                that can be used by other components to show or close the modal.
```
Comment (lnr0626) [inline lib/surface_site_web/live/events.ex:217]:
```suggestion
                that can be used by other components to open and close the modal.
```

**Review by lnr0626 — COMMENTED** [review]
I had a few suggestions, great work - loving the additional examples and more detailed docs

**Comment by paulstatezny** [comment]
@msaraiva Thanks for tagging me! I'll take a look at this in the next 36 hours at most.

File: lib/surface_site_web/live/events.ex [hunk 11]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+                See the properties <LiveRedirect label="Event" to="/properties#event-property" /> section for more details about event properties.
+
+              <#Markdown>
+
+                ### Stateless component
+                The simplest case you need to pass an event is when you create a stateless component that includes an element
+                that defines a server binding (event). Since the component is stateless, it cannot handle the event by itself
+                so it needs to receive the event handler as a property.
+
+                For example, imagine a `Button` stateless component that raises an event when the user clicks on it.
```
Comment (paulstatezny) [inline lib/surface_site_web/live/events.ex:142]: What's the most precise verb to use here for dispatching/emitting/triggering an event? I think there might be something more fitting than `raise` but I'm not sure.

Or maybe it doesn't matter. 🙂

File: lib/surface_site_web/live/events.ex [hunk 12]
```diff
@@ -92,29 +95,174 @@ defmodule SurfaceSiteWeb.Events do
...
+
                 Another great thing about Surface's approach is that it makes passing events as properties
                 also more intuitive. Using phoenix templates, unless you always pass both, the event and the target,
                 you cannot be sure where the event will be handled. You need to know upfront if there's a `phx-target`
-                defined for that DOM element inside that component.
+                defined for that DOM element inside that component. Using Surface, the event is **always** passed
+                along with the related target, assuming, by default, that the target is the caller component/view.
+
+                In the above examples the events have been handled by the component itself.
+                Sometime the component does not know how to handle an event and you would like to handle this event in the parent component.
```
Comment (paulstatezny) [inline lib/surface_site_web/live/events.ex:128]:
```suggestion
                Sometimes the event needs to be handled by the parent component.
```

**Comment by paulstatezny** [comment]
It's great to dive into more concrete examples and highlight details in the docs. 👍 I think there is some room for tightening up the voicing in some of these and using more succinct or clear verbiage; I left a couple suggestions. But that can be iterated on, too. Thanks!

File: lib/surface_site_web/live/events.ex (hunk 11 above)
Comment (msaraiva) [reply lib/surface_site_web/live/events.ex:142]: I guess triggering is the way to go. Would that sound better?

```
For example, imagine a stateless `Button` component that triggers an event when clicked.
```

**Review by msaraiva — APPROVED** [review]

**Pull request merged**